### ISSUE TITLE
iOS Mobile Renderer Black Screen

### ISSUE THREAD

**Issue by TCROC** [godotengine/godot#99682]
### Tested versions

- Reproduceable In: 4.4.dev.mono.custom_build.f0e971730

### System information

Godot v4.4.dev.mono (f0e971730) - macOS 15.0.1 - Multi-window, 1 monitor - Metal (Forward+) - integrated Apple M2 (Apple8) - Apple M2 (8 threads)

### Issue description

When exporting to iOS set to use the mobile renderer, it renders with a black screen:

![Image](https://github.com/user-attachments/assets/b788e0ba-b23a-4b56-9e96-db3272d29f4f)

When exporting to iOS set to use the Forward+ renderer, it renders just fine:

![Image](https://github.com/user-attachments/assets/9e531044-4d5c-46b6-8ac8-eb84621516c8)

Here are the respective logs:

### Note

The Forward+ Renderer Log has many errors, but that is the one that actually renders correctly.  The Mobile Renderer log only has 1 error, but renders with a black screen.

### Mobile Renderer Log

```
Godot Engine v4.4.dev.mono.custom_build.f0e971730 (2024-11-16 22:16:24 UTC) - https://godotengine.org
Metal 3.1 - Forward Mobile - Using Device #0: Apple - Apple A12 GPU (Apple5)

ERROR: Mouse is not supported by this display server.
at: mouse_get_position (servers/display_server.cpp:508)
```

### Forward+ Renderer Log

[forward_plus_log.txt](https://github.com/user-attachments/files/17905110/forward_plus_log.txt)

### Steps to reproduce

1. Use Godot with double precision and mono support.
    - I'm not certain if this is necessary as I haven't tested other builds, but that is what I am currently using.
2. Open the MRP
3. Export to iOS
4. Install to an iPhone
5. See that a black screen renders
6. Change project settings mobile rendering method to Forward+
7. Export to IOS
8. Install to an iPhone
9. See that cube renders

### Minimal reproduction project (MRP)

[mrp.zip](https://github.com/user-attachments/files/17905053/mrp.zip)

Edit:

Here's the table of tested snapshots:

✅ = ~vulkan~ mobile works as expected
❌ = ~vulkan~ mobile blank screen

| Version  | Status  |
| -------- | ------- |
| 4.3.stable  | ✅ |
| 4.4-dev1 | ❌    |
| 4.4-dev2 | ?     |
| 4.4-dev3 | ?     |
| 4.4-dev4 | ?     |
| 4.4-dev5 | ❌    |

Edit:

I crossed out ~vulkan~ above.  @bruvzg pointed out [here](https://github.com/godotengine/godot/issues/99682#issuecomment-2500088341) that we are actually defaulting to the metal rendering device on ios now.  Changing it back to use Vulkan in the project settings as described in that comment makes the MRP work.  So it appears to be specifically related to the metal driver introduced in this pr: https://github.com/godotengine/godot/pull/88199

**Comment by Calinou** [comment]
Can you reproduce this on 4.3.stable?

**Comment by TCROC** [comment]
I will go check

**Comment by TCROC** [comment]
Also note that I reproduced this with normal Godot.  It occurred in the 4.4-dev5 snapshot without mono support.  And I believe those snapshots do not have double precision either right?  Which would lead me to believe this issue is not related to mono support or double precision.

**Comment by TCROC** [comment]
| Version  | Status  |
| -------- | ------- |
| 4.3.stable  | ✅ |
| 4.4-dev1 | ❌    |
| 4.4-dev2 | ?     |
| 4.4-dev3 | ?     |
| 4.4-dev4 | ?     |
| 4.4-dev5 | ❌    |

So it was first introduced somewhere between 4.3.stable and 4.4-dev1.

@Calinou I found where it was introduced in regards to snapshots :).

**Comment by bruvzg** [comment]
> Metal 3.1 - Forward Mobile - Using Device #0: Apple - Apple A12 GPU (Apple5)

Default backend for Mobile/Forward+ on macOS and iOS was changed to new Metal renderer. While Vulkan renderer should support A12 (it is the oldest SOC it should run on), I'm not sure what's the minimum requirement for Metal renderer. Try setting `rendering/rendering_device/driver.ios` in the project settings to `vulkan`.

**Comment by TCROC** [comment]
@bruvzg I will give this a shot

@Calinou I bisected and found the PR that introduced this issue.  Looks like it @stuartcarnie 's pr here: https://github.com/godotengine/godot/pull/88199

Thank you btw @stuartcarnie for the metal support!  I love getting as close to the native drivers for target platforms as possible! :). I just believe I may have found an issue with this PR that introduces black screen on ios.

**Comment by TCROC** [comment]
@bruvzg I can confirm that changing the ``rendering_device/driver.ios`` back to ``vulkan`` fixes this :)

**Comment by bruvzg** [comment]
For the reference, Metal render seems to be working fine on A14 (10. gen. iPad) and M4 Pro. Unlike A12, these devices have `Metal 3.2` printed when starting, not `Metal 3.1`. So this might be the requirement.

```
Metal 3.1 - Forward Mobile - Using Device #0: Apple - Apple A12 GPU (Apple5)

Metal 3.2 - Forward Mobile - Using Device #0: Apple - Apple A14 GPU (Apple7)
Metal 3.2 - Forward+ - Using Device #0: Apple - Apple M4 Pro (Apple9)
```

[Officially supported Metal 3 devices](https://support.apple.com/en-us/102894) are A12X/A13+, so base A12 might be too old and have incomplete Metal implementation. We probably should update documentation with new min. system requirements for Metal renderer.

**Comment by TCROC** [comment]
@bruvzg Should we have it automatically fall back to Vulkan if we detect at startup that it is the old one?

Also something to note, Forward+ works without issue.  I was only having issues with the Mobile renderer

**Comment by bruvzg** [comment]
> Should we have it automatically fall back to Vulkan if we detect at startup that it is the old one?

Yes, but currently fallback only works if a specific renderer is completely missing, in this case renderer init works, but shader compiler fails.

> Also something to note, Forward+ works without issue. I was only having issues with the Mobile renderer

In this case, it probably is a Metal renderer bug, not a hardware limitation.

**Comment by stuartcarnie** [comment]
I tried the 2D lights and shadows demo, as I already had that setup, and it is still working with the Forward Mobile renderer, but Forward+ is definitely returning a lot of errors, so I don't believe it is working without issue. I'll try your MRP next.

**Comment by TCROC** [comment]
For some extra info, my phone is:

```
iOS Version: 17.6.1
Model Name: iPhone XR
Model Number: MT362LL/A
```

**Comment by stuartcarnie** [comment]
> > Also something to note, Forward+ works without issue. I was only having issues with the Mobile renderer
> 
> In this case, it probably is a Metal renderer bug, not a hardware limitation.

Forward+ is not working correctly, as many of the pipelines are failing to compile, which are in the logs @TCROC included. I was able to replicate that on my iPhone 13. I'll have to take a look and see why it's failing, as I expect it _should_ work.

I'm mostly interested in why the mobile renderer isn't working for you, as there are no logs and no errors / warnings suggesting it is even compiling anything.

**Comment by TCROC** [comment]
> > > Also something to note, Forward+ works without issue. I was only having issues with the Mobile renderer
> > 
> > 
> > In this case, it probably is a Metal renderer bug, not a hardware limitation.
> 
> Forward+ is not working correctly, as many of the pipelines are failing to compile, which are in the logs [@TCROC](https://github.com/TCROC) included. I was able to replicate that on my iPhone 13. I'll have to take a look and see why it's failing, as I expect it _should_ work.
> 
> I'm mostly interested in why the mobile renderer isn't working for you, as there are no logs and no errors / warnings suggesting it is even compiling anything.

Did the MRP work on your iPhone 13 with mobile renderer?

**Comment by stuartcarnie** [comment]
> Did the MRP work on your iPhone 13 with mobile renderer?

Just trying that now

**Comment by stuartcarnie** [comment]
It is reproducible – the issue is in the Metal renderer and related to how SPIRV-Cross generates the shader interface for iOS vs macOS for certain shaders. Turns out that the mobile renderer is also partially rendering (like Forward+), a full-screen quad is rendering black for mobile.

**Comment by TCROC** [comment]
Perfect! Thank you for reproducing @stuartcarnie! And happy Thanksgiving if you celebrate it! :) I guess we can update the tags now that it has been confirmed :)

**Comment by stuartcarnie** [comment]
Fix coming via #99820

**Comment by TCROC** [comment]
@stuartcarnie I'm not saying you are "the man". But if the "the man" were to hypothetically exist, he would likely look something like you 😎

I'll try out your PR in my fork in the morning! 🔥

**Comment by stuartcarnie** [comment]
@TCROC I think I see that you have an Apple5 GPU device (2018), which is unfortunately too low for the change I had to make. The minimum version is now an Apple6 family GPU, which is from 2019. SPIRV-Cross has some issues with mixing Apple GPU features between Apple5 and Apple6 era GPUs, which directly affect the Metal renderer 😔 

I'm hoping we can support some sort of automatic fall-back to MoltenVK.

**Comment by TCROC** [comment]
@stuartcarnie Perfect!  Let me know what solution  you come to and I'll gladly test it out :)  Thank you! You are the best! :)

**Comment by TCROC** [comment]
@maxelusgit try changing your rendering driver to vulkan:

![Image](https://github.com/user-attachments/assets/7961ab4f-fd76-477b-af6a-e8de27217335)

**Comment by stuartcarnie** [comment]
@TCROC #99820 will now support your iPhone

**Comment by TCROC** [comment]
@stuartcarnie perfect! I'll try that tomorrow :)

**Comment by stuartcarnie** [comment]
@maxelusgit do you have a MRP we can look at, as that will help us troubleshoot the problem. I have an iPhone 12 Pro, so older than yours.

**Comment by stuartcarnie** [comment]
@TCROC were you able to try this latest PR on your A12 device (Apple5 GPU)?

**Comment by TCROC** [comment]
@stuartcarnie not yet. I had some things come up that I had to deal with. I try to get around to testing before end of week

**Comment by iusn** [comment]
Godot Engine mono v4.4.dev7.official.46c8f8c5c - https://godotengine.org
Metal 3.2 - Forward Mobile - Using Device #0: Apple - Apple A18 Pro GPU (Apple9)

exporting to iOS set to use the mobile renderer, it renders with a black screen. like this issue.

**Comment by stuartcarnie** [comment]
> [!IMPORTANT]
>
> @iusn, et al, the fix, #99820 did not make the dev-7 release
